### PR TITLE
Particle system optimizations

### PR DESCRIPTION
**Pull request by willeastcott**
I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/framework/components/particle-system/component.js [hunk 1]
```diff
@@ -574,10 +574,69 @@ Object.assign(pc, function () {
...
+                }
+                if (numSteps) {
+                    numSteps = Math.min(numSteps, emitter.maxSubSteps);
+                    for (i = 0; i < numSteps; i++) {
+                        emitter.addTime(emitter.fixedTimeStep, false);
+                    }
+                    var stats = this.system.app.stats.particles;
+                    stats._updatesPerFrame += numSteps;
+                    stats._frameTime += emitter._addTimeTime;
+                    emitter._addTimeTime = 0;
```
Comment (yaustar) [inline src/framework/components/particle-system/component.js:631]: Does this need to be wrapped in `#ifdef PROFILER`?

File: src/framework/components/particle-system/component.js [hunk 2]
```diff
@@ -696,27 +755,43 @@ Object.assign(pc, function () {
...
-            if (this.system.app.scene.layers) {
-                this.system.app.scene.layers.off("add", this.onLayerAdded, this);
-                this.system.app.scene.layers.off("remove", this.onLayerRemoved, this);
+            var data = this.data;
+            var system = this.system;
+
+            // Remove from enabled particle systems array
+            var idx = system._enabledParticleSystems.indexOf(this);
+            if (idx > -1) {
+                system._enabledParticleSystems.splice(idx, 1);
```
Comment (yaustar) [inline src/framework/components/particle-system/component.js:780]: Minor: I would prefer this to be done through a function/property as `_enabledParticleSystems` is considered private to the `system`. But if we do this elsewhere, that's fine.

File: src/framework/components/particle-system/component.js [hunk 3]
```diff
@@ -574,10 +574,69 @@ Object.assign(pc, function () {
             }
         },
 
+        _update: function (dt) {
+            var emitter = this.emitter;
+            if (!emitter.meshInstance.visible) return;
+
```
Comment (yaustar) [inline src/framework/components/particle-system/component.js:580]: If the entity/component is disabled, is `emitter.meshInstance.visible` false anyway and therefore the emitter is not updated?

**Review by raytranuk — COMMENTED** [review]
A description of what this optimizes and high level as to how, would useful - from what I gather: iteration per particle is done per emitter instead of per system - and I also see some code tidy up. Profiling numbers before and after would be great too.

**Comment by willeastcott** [comment]
This has gone a bit stale so I'm closing it (for now).

**Pull request closed without merging**